src/wasm/wasm-s-parser.cpp

```diff
@@ -579,6 +579,11 @@ size_t SExpressionWasmBuilder::parseTypeUse(Element& s,
     paramOrResultExists = true;
     result = parseResult(*s[i++]);
   }
+  // If none of type/param/result exists, this is equivalent to a type that does
+  // not have parameters and returns nothing.
+  if (!typeExists && !paramOrResultExists) {
+    paramOrResultExists = true;
```

Comment:
this confuses me. it reads as "if no type and no param, then set param-or-result to true"? that seems like the opposite of what I'd expect?

Reply:
If we have none of `(type)`, `(param)`, or `(result)`, that's not invalid, because that's equivalent to `(param)`/`(result)` with void return type and no parameters. This variable is to figure out which of the typeuse, between `(param)`/`(result)` and `(type)`, is specified, and if we don't have anything, that means we have `(param)`/`(result)` (with void return type and no param types)

Reply:
Thanks, I think I see. Would it maybe be simpler to not track `paramOrResultExists`? And have only `typeExists`?

Reply:
It's hard, because we have to error out when both (type) and (param)/(result) exist and they don't match. If only (type) is specified and (param)/(result) is not, this is not an error: we fill in param/result based on the (type). But if (type) is given and also (param)/(result) is given, even if they are empty, we have to check them to see if they match.

So in short, if there is no (param)/(result), it is considered to be given (as empty param/results) only if there is no (type). If there's a (type), we should distinguish cases when (param)/(result) is not given vs. they are given as empty lists (`(param)` or `(result)`, without types).

Reply:
I see, thanks. Well, sounds kind of complicated, but yeah, unavoidable I guess...